feat(main): expand file ~\ or ~/ prefix on Windows

rkitover opened the pull request:
In command_line_scan() for MSWIN, expand `"~\"` or `"~/"` prefixed paths to
the USERPROFILE environment variable for the user's profile directory.

Fix #23901

rkitover commented on src/nvim/path.c:
```diff
@@ -1872,7 +1873,13 @@ char *fix_fname(const char *fname)
 #else
-  if (!vim_isAbsName(fname)
+  if (*fname == '~' && (fname[1] == '/' || fname[1] == '\\')) {
+    const char *profile_dir = os_getenv("USERPROFILE");
```
That function is `static`, how do I get `homedir` out of `env.c`?

rkitover commented:
> What _problem_ is this solving? Give a specific use-case, describing the steps exactly.

So, PowerShell on Windows recently removed the feature where arguments such as `~/.ssh/config` or `~/.vimrc` are expanded to `${env:USERPROFILE}/.ssh/config` etc..

What this means is, in PowerShell, when you run a command such as:

```powershell
nvim ~/.ssh/config
```
, what happens is that nvim tries to open `${PWD}/~/.ssh/config` because that looks like a relative path to a literal `~` directory to it.

rkitover commented:
I don't understand the lint failure, what am I doing wrong there?

justinmk commented:
> nvim (gi ~/.ssh/config)

`nvim "+n ~/.ssh/config"` is another workaround. https://github.com/neovim/neovim/issues/23901#issuecomment-1575551962

We might want to address this, but need to be sure that `fix_fname` is the right place to do it. (I haven't analyzed the codepaths involved.)

rkitover commented:
> We might want to address this, but need to be sure that fix_fname is the right place to do it. (I haven't analyzed the codepaths involved.)

So I tried to do this in `os_open()` in `os/fs.c` first, but it is too late to do it there because `fix_fname()` has already converted the path to an absolute path to a relative `~` directory, this is why I put it there.

> I had heard differently in another issue. Maybe it's configurable? Can you provide a reference (which should have been mentioned from the start, in the PR description)

There used to be an experimental feature for this in `get-experimentalfeature` that you could enable, I don't remember which exactly, but it has been removed.

In the unreleased development version of PowerShell, which may not be released for quite a while yet, there is an experimental feature for this, `PSNativeWindowsTildeExpansion`, see [here](https://github.com/PowerShell/PowerShell/blob/master/experimental-feature-windows.json).

rkitover commented:
I believe it would be better if we address this in neovim, because this PowerShell experimental feature will not be available for a while yet, and when it is it would require enabling it, and if this is done in neovim it would also work in WinPS (`powershell.exe`) and `cmd.exe`.

justinmk commented:
If there is not already an issue tracking this, please consider opening an issue with your notes.

rkitover commented:
Done, the [issue](https://github.com/neovim/neovim/issues/23901) you linked was it.

rkitover commented:
@justinmk ping, any planned action on this PR? I do believe I put the code in the right place and including this feature is justified, and I would really like it to be released for Windows users.

dundargoc commented:
For starters the CI is failing, you need to run `make format` to format your code correctly.

zeertzjq reviewed (dismissed):
I don't think this is the correct place to do this expansion. I think this should be done in main.c, after the `char *p = xstrdup(argv[0])` line.

rkitover commented:
Thanks guys, I'm on a road trip right now I will make these changes when I get back tonight.

rkitover commented:
By the way, the format target is broken on Windows/MSVC, I'll try to look at that later.

rkitover commented:
@zeertzjq

I looked at that section of code and that does look like the right place to put this expansion to fix the problem I am trying to solve. But I wonder, do we want this expansion for only file arguments or for any file handling in general, e.g. from vimscript, ex commands, etc.. I think the latter behavior would be nicer, because then I can do e.g. `:e ~/.ssh/config` and it would work as well.

Your thoughts please.

zeertzjq commented:
> @zeertzjq
> 
> I looked at that section of code and that does look like the right place to put this expansion to fix the problem I am trying to solve. But I wonder, do we want this expansion for only file arguments or for any file handling in general, e.g. from vimscript, ex commands, etc.. I think the latter behavior would be nicer, because then I can do e.g. `:e ~/.ssh/config` and it would work as well.
> 
> Your thoughts please.

For Ex commands this is already done in `expand_env_esc()`.

rkitover commented:
Sorry I should have tested this before posting, that does indeed work. However, `:echo readfile("~/.ssh/config")` does not work, and it does not work in my current implementation either.

rkitover commented:
@zeertzjq 

And in fact, `:echo readfile("~/.ssh/config")` does not work on UNIX either.

I'd like to propose a solution for full support for tilde expansion on Windows and UNIX for all neovim file operations. I think I know how to do this now.

- Fix `vim_isAbsName()` in `path.c` to return true for paths with a tilde expansion, because they are in fact absolute paths, this way `fix_fname()` in `path.c` will not "fix" them.
- Perform the expansion in `os_fopen()` in `os/fs.c` using a helper function on both UNIX and Windows.

I want to handle `~username/` expansions as well, this may be of use to admin users. This is much harder to do on Windows and I may not be able to fully support it, but I will try for at least partial support.

If you don't like this idea I will go with your solution of putting the expansion in `main.c` because that does in fact fix the main problem I am trying to fix.

zeertzjq commented:
> I'd like to propose a solution for full support for tilde expansion on Windows and UNIX for all neovim file operations.

I don't think this should be supported.

rkitover commented:
@zeertzjq How's this?

zeertzjq commented on runtime/doc/news.txt:
```diff
@@ -18,6 +18,12 @@ BREAKING CHANGES IN HEAD                                    *news-breaking-dev*
 The following changes to UNRELEASED features were made during the development
 cycle (Nvim HEAD, the "master" branch).
 
+MAIN
+
+* On Windows, filename arguments on the command-line prefixed with "~/" or
+  "~\" are now expanded to the user's profile directory, not a relative path
+  to a literal "~" directory.
+
```
This is the wrong place to put this. It should be in the "EDITOR" subsection of the "NEW FEATURES" section.

zeertzjq commented on src/nvim/main.c:
```diff
@@ -1444,6 +1444,18 @@ static void command_line_scan(mparm_T *parmp)
+      // On Windows expand "~\" prefix in file names to profile directory.
+#ifdef MSWIN
+      if (*p == '~' && (p[1] == '/' || p[1] == '\\')) {
+        const char *profile_dir = os_getenv("USERPROFILE");
```
Need to check if `os_getenv()` returns `NULL`

zeertzjq commented on src/nvim/main.c:
```diff
@@ -1444,6 +1444,20 @@ static void command_line_scan(mparm_T *parmp)
+      // On Windows expand "~\" or "~/" prefix in file names to profile directory.
+#ifdef MSWIN
+      if (*p == '~' && (p[1] == '\\' || p[1] == '/')) {
+        const char *profile_dir = os_getenv("USERPROFILE");
+        if (profile_dir) {
```
On second thought, it may be better to use `vim_getenv("HOME")` here. That'll return the `homedir` variable in os/env.c, and there is no need to check for `NULL`.

rkitover commented:
@zeertzjq done.

zeertzjq commented on src/nvim/main.c:
```diff
@@ -1444,6 +1444,18 @@ static void command_line_scan(mparm_T *parmp)
+        size_t size = strlen(profile_dir) + strlen(p);
+        char *tilde_expanded = xmalloc(size);
+        snprintf(tilde_expanded, size, "%s%s", profile_dir, p + 1);
+        xfree(p);
```
`profile_dir` is allocated, so it needs to be freed as well.
```suggestion
        xfree(p);
        xfree(profile_dir);
```

rkitover commented:
@zeertzjq sorry, fixed.

rkitover committed:
feat(main): expand file ~\ or ~/ prefix on Windows

In command_line_scan() for MSWIN, expand "~\" or "~/" prefixed paths to
the USERPROFILE environment variable for the user's profile directory.

Fix #23901

Signed-off-by: Rafael Kitover <rkitover@gmail.com>

rkitover commented:
@zeertzjq @justinmk @dundargoc

Really appreciate your help with this, thank you.

Looking forward to this being released so I can tell my PowerShell friends about it.

I am going to try to do another PR here for the format target on Windows.

I am going to try to get this feature into vim as well.

justinmk commented on src/nvim/main.c:
```diff
@@ -1444,6 +1444,19 @@ static void command_line_scan(mparm_T *parmp)
+      // On Windows expand "~\" or "~/" prefix in file names to profile directory.
+#ifdef MSWIN
+      if (*p == '~' && (p[1] == '\\' || p[1] == '/')) {
+        char *profile_dir = vim_getenv("HOME");
```
Does this happen before `init_homedir`, or why not use `os_homedir` ? $HOME is not always available Windows.

Also needs test coverage.

But I'm confused about why this is needed, because https://github.com/neovim/neovim/issues/23901#issuecomment-2033479836 indicated that powershell would provide this feature without every program needing to implement it manually.

rkitover commented:
> Does this happen before init_homedir, or why not use os_homedir ? $HOME is not always available Windows.

`$HOME` should always be available on Windows, at least in any normal situation. `neovim` would probably not function correctly at all in that case.

If it does happen that it is not available in some situation we can deal with that separately, I think.

This code is in `os/env.c`, `os_homedir()` gets the homedir from libuv, `init_homedir()` sets `homedir` from `os_homedir()` while `vim_getenv("HOME")` returns `homedir`. So this is essentially the same thing.

I should mention that I did test this and it does work as intended.

> Also needs test coverage.

I will do a separate PR for that.

> But I'm confused about why this is needed, because https://github.com/neovim/neovim/issues/23901#issuecomment-2033479836 indicated that powershell would provide this feature without every program needing to implement it manually.

It is currently a major usability issue.

For many PowerShell users, `nvim` is the native program they will most often call with tilde paths.

Someone mentioned that tab completion for tilde paths may be backported to the full expansion behavior, but it is unknown whether and when that will happen. And that would still require tab expansion.

This will also work in WinPS (`powershell.exe`, version 5) which will not be updated for a very long time to come. It does expand tilde paths with tab, but this will work without expansion.

It may also be helpful for people using other shells, I don't know.

The experimental feature for handling tilde path arguments to native programs in the development branch of PowerShell will most likely not be available for months to come, and even then it may only be available as an experimental feature that requires it to be enabled manually.

Given that this is not a large change, and does not break anything, I think it's justified, given that it is a major usability issue for Windows users.

You may consider PowerShell not handling tilde paths itself like a UNIX shell absolutely ridiculous, and I agree, but that is the situation we are in.

justinmk commented:
> `$HOME` should always be available on Windows, at least in any normal situation.

Since when? https://superuser.com/a/607110

> Neovim would probably not function correctly at all in that case.

Completey wrong. `os_homedir` existst to deal with the lack of $HOME, at least on older(?) Windows versions.



This does not inspire me with confidence in this PR.

zeertzjq commented:
@justinmk Please read the source code of `vim_getenv()`.

justinmk commented:
ok, we special case $HOME internally. would os_homedir be clearer?

rkitover commented:
> ok, we special case $HOME internally. would os_homedir be clearer?

I think so, but it is a minor stylistic difference, I will make this change in my subsequent PR for the test that I promised you. Please give me a couple weeks, things are difficult for me right now.

I also promised to fix the format target on Windows and I will try to do that.
